wingman.el

```diff
@@ -1024,5 +1113,230 @@ filtering is performed."
+
+    (wingman--log 2 "Sending completion request to %s via gptel-request (%s)"
+                  backend-name model)
+
+    (wingman--log 4 "Full prompt (%d chars): %s\n\n"
+                  (length prompt)
+                  (if (> (length prompt) 1000)
+                      (concat (substring prompt 0 500)
+                              "\n[... truncated "
+                              (number-to-string (- (length prompt) 1000))
+                              " chars ...]\n"
+                              (substring prompt -500))
+                    prompt))
+
+    (unless backend-object
+      (user-error "Wingman: Unknown backend '%s' for completion request" backend-name))
+
+    (gptel-request prompt
+      :buffer (current-buffer)
+      :position start-marker
+      :callback
+      (lambda (response info)
+        (let ((origin-buffer (plist-get info :buffer)))
+          (with-current-buffer origin-buffer
+            (goto-char (plist-get info :position))
+            (cond
+             ((stringp response)
+              (let ((cleaned-response (string-trim response)))
+                ;; Remove markdown code fences if they exist.
+                (when (string-match "^```\\(?:\\w+\\)?\n\\(.*\\)\n```$" cleaned-response)
+                  (setq cleaned-response (match-string 1 cleaned-response)))
+
+                (wingman--gptel-render cleaned-response origin-buffer)
+                (message "Wingman: Completion available. Use TAB/S-TAB to accept.")))
+
+             ((eq response 'abort)
+              (message "Wingman: Completion aborted"))
+
+             ((null response)
+              (let ((status (plist-get info :status))
+                    (error-data (plist-get info :error)))
+                (message "Debug: HTTP Status: %s" status)
+                (message "Debug: Error data: %s" error-data)
+                (message "Debug: Full info: %s" info)
+                (message "Wingman: Completion failed - %s" (or status "Unknown error"))))
+
+             (t
+              (message "Debug: Unexpected response type: %s" response)
+              (message "Wingman: Unexpected response type"))))))
+
+      :context (list :inline-completion t))
+
+    (message "Wingman: Requesting completion from %s (%s)..."
+             backend-name model)))
+
+(defun wingman--build-emulated-fim-prompt (ctx)
+  "Build the prompt string for gptel from wingman context CTX."
+  (let* ((pre (alist-get 'prefix ctx))
+         (mid (alist-get 'middle ctx))
+         (suf (alist-get 'suffix ctx))
+         (extra-ctx (wingman--extra-context ctx))
+         (lang (wingman--lang-from-major-mode major-mode))
+         (context-section
+          (if (and extra-ctx (> (length extra-ctx) 0))
+              (let ((context-chunks
+                     (mapcar (lambda (chunk)
+                               (format "--- From: `%s` ---\n```\n%s\n```"
+                                       (alist-get 'filename chunk)
```

Comment:
Since this is going to third party providers, by default consider not using the full file path. It may leak the user's full name. `abbreviate-file-name` is an option to substitute the home directory with `~`. Another option would be to use the project-relative path if in a project.